## CS 156 HW 5
### Ankush Hommerich-Dutt

1. **[c]**

E_in = $\sigma^2(1 - \frac{d+1}{N})$  

If we plug in $\sigma = 0.1, d = 8$ and E_in > 0.008, we get:

$(0.1)^2(1 - \frac{9}{N}) > 0.008$. Solving this for $N$:

$1 - \frac{9}{N} > 0.8 \longrightarrow N > \frac{9}{0.8} = 45$

So $N$ has to be bigger than 45.

Among the answer choices, 100 is the smallest number that is bigger than 45.

2. **[d]**

We see that for large $x_1$, we see that the classification is -1, and for large $x_2$, the classification is +1. Therefore, if $w_1$ is negative, then for large $x_1$, we will get that $w^Tx$ is negative, and if $w_2$ is positive, then for large $x_2$, we will get that $w^Tx$ is positive.

This corresponds to answer choice d, where $w_1 < 0; w_2 > 0$

(Of course all of the w should have a tilde above them)

3. **[d]**

We know that a linear model in a $d$ dimensional space has a VC dimension of $d + 1$ (this was proven in class for the perceptron). Therefore, in the transformed space, the dimension is 15, and so the linear model would have a VC dimension of 16. 20 is the smallest number amongst the answer choices that is larger than 16.

4. **[e]**

We can use the chain rule to take the partial derivative. We see that it is answer choice e. I am not sure what work to show for this, as it is a one step process. The 2 goes to the front, we keep our original function that was in parenthesis, and then we multiply it by the derivative of that function with respect to (wrt) u. The derivative of $ue^v$ wrt $u$ is $e^v$, and the derivative of $-2ve^{-u}$ wrt $u$ is $2ve^{-u}$. This all corresponds to answer choice e.

In [25]:
import random
import math
import numpy as np
from numpy import linalg as LA
import matplotlib.pyplot as plt

def error_func(u, v):
    return (u*np.exp(v) - 2*v*np.exp(-u))**2

def error_gradient(u, v):
    return np.array([(u*np.exp(v) - 2*v*np.exp(-u))*(np.exp(v) + 2*v*np.exp(-u))*2, \
                     (u*np.exp(v)  - 2*v*np.exp(-u))*(u * np.exp(v) - 2*np.exp(-u))*2])


def converge(weights, error, eta):
    num_iter = 0
    while error_func(weights[0], weights[1]) > error:
        weights = weights - (eta * error_gradient(weights[0], weights[1]))
        num_iter += 1 
    return weights, num_iter      

def coordinate_descent(weights, iterations, eta):
    for i in range(iterations):
        error = error_gradient(weights[0], weights[1]) 
        weights = weights - (eta * np.array([error[0], 0]))
        
        error = error_gradient(weights[0], weights[1]) 
        weights = weights - (eta * np.array([0, error[1]]))
    
    return error_func(weights[0], weights[1])

    
    
def main():
    weights = np.array([1,1])
    new_weights, iters = converge(weights, 10**(-14), 0.1)
    print("number of iterations: ", iters)
    print("final u and v: ", new_weights)
    error_descent = coordinate_descent(weights, 15, 0.1)
    print("the error for coordinate descent: ", error_descent)
                          
main() 
    


number of iterations:  10
final u and v:  [0.04473629 0.02395871]
the error for coordinate descent:  0.13981379199615315


5. **[d]**

As we can see, it took 10 iterations to converge to an error below $10^{-14}$

6. **[e]**

After the convergence based on the criteria above, we can see our final $u$ is 0.045 and our final $v$ is 0.024. This corresponds to answer choice e

7. **[a]**

After 15 full iterations of coordinate descent, we see that the error is 0.14, which corresponds closest to answer choice a of $10^{-1}$

In [26]:
import random
import numpy as np
from numpy import linalg as LA

class Point:
    def __init__(self, x, y):
        self.val = np.array([1, x, y])
        
    def outcome(self, line):
        target = line.evaluate(self.val[1])
        if self.val[2] > target:
            return 1
        return -1
        
class Line:
    def __init__(self, point1, point2):
        self.slope = (point2.val[2] - point1.val[2]) / (point2.val[1] - point1.val[1])
        self.intercept = (self.slope * (-point1.val[1])) + point1.val[2]
        
    def evaluate(self, x):
        return (self.slope * x) + self.intercept
    
        
def random_point():
    return Point(random.uniform(-1,1), random.uniform(-1,1))

def target_function():
    point1 = random_point()
    point2 = random_point()
    return Line(point1, point2)


def iteration(outcomes, weight_vec, eta, n):
    old_weight = weight_vec
    perms = np.random.permutation(n)
    for i in perms:
        grad = -(outcomes[i][0].val * outcomes[i][1]) / \
        (1 + np.exp(weight_vec.dot(outcomes[i][0].val) * outcomes[i][1]))
        weight_vec = weight_vec - (eta * grad)
    
    if LA.norm(weight_vec - old_weight) < 0.01:
        return weight_vec, True
    return weight_vec, False

    
def main(n):
    target = target_function()
    data = [random_point() for x in range(n)]
    outcomes =  list(map(lambda x: (x,x.outcome(target)), data))
    
    eta = 0.01
    converge = False 
    num_epochs = 0
    w = np.array([0,0,0])
    while converge != True:
        w, converge = iteration(outcomes, w, eta, n)
        num_epochs += 1  
        
    new_data = [random_point() for x in range(1000)]
    new_outcomes =  list(map(lambda x: (x, x.outcome(target)), new_data))
    E_out_sum = 0
    for i in range(1000):
        E_out_sum += np.log(1 + np.exp(w.dot(new_outcomes[i][0].val) * new_outcomes[i][1] * -1))
    E_out = E_out_sum / 1000
    return E_out, num_epochs

test = [main(100) for i in range(100)]
print("N = 100: ")
print("E_out: ", sum(i for i, _ in test) / 10w0)
print("num_epochs: ", sum(i for _, i in test) / 100)


N = 100: 
E_out:  0.10132224456326522
num_epochs:  337.59


8. **[d]**

We get an E_out of 0.101 with N = 100, which corresponds closest to answer choice d.

9. **[a]**

We converged after 337 epochs on average, which corresponds to answer choice a of 350.

10. **[b]**

We see that in the PLA, the weight vector after picking one point is updated by $w = w + x_ny_n$, and so $x_ny_n$ needs to equal -$\nabla e_n$ to draw an analogy to stochastic gradient descent (we can ignore eta since it is a constant). If we take the gradient of answer choice b, we get $\nabla e_n = -x_ny_n$, which is exactly what we needed. 